### PR TITLE
auditor and issuer are the same node #283

### PR DESCRIPTION
**Pull request by adecaro**
Signed-off-by: Angelo De Caro <adc@zurich.ibm.com>

### CODE REVIEW

File: token/services/ttx/support.go [hunk 1]
```diff
@@ -54,3 +57,128 @@ func StoreEnvelope(context view.Context, tx *Transaction) error {
...
+		Payload:      payload,
+	}
+	return nil
+}
+
+func (s *localSession) Receive() <-chan *view.Message {
+	return s.readChannel
+}
+
+func (s *localSession) Close() {
```
Comment (mbrandenburger) [inline token/services/ttx/support.go:182]: is this somehow used? For instances as check if Send Or Receive is called?

File: integration/nwo/token/common/ppmgen.go [hunk 2]
```diff
@@ -10,7 +10,8 @@ import (
 	"io/ioutil"
 	"path/filepath"
 	"strconv"
-	"strings"
+
+	msp2 "github.com/hyperledger-labs/fabric-token-sdk/token/core/identity/msp"
 
 	msp "github.com/IBM/idemix"
 	math3 "github.com/IBM/mathlib"
```
[the remark this answers is not in the document]
Comment (adecaro) [reply integration/nwo/token/common/ppmgen.go:17]: okay

File: token/services/ttx/support.go [hunk 3]
```diff
@@ -54,3 +57,128 @@ func StoreEnvelope(context view.Context, tx *Transaction) error {
...
+}
+
+// NewLocalBidirectionalChannel creates a new bidirectional channel
+func NewLocalBidirectionalChannel(caller string, contextID string, endpoint string, pkid []byte) (*LocalBidirectionalChannel, error) {
+	ID, err := comm.GetRandomNonce()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate session ID")
+	}
+	lr := make(chan *view.Message, 10)
+	rl := make(chan *view.Message, 10)
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/services/ttx/support.go:95]: okay

File: token/services/ttx/support.go (hunk 1 above)
Comment (adecaro) [reply token/services/ttx/support.go:182]: let me update the code to check that flag.

**Comment by adecaro** [comment]
Thanks @mbrandenburger , I have addressed your comments and rebased :)

File: token/services/ttx/support.go [hunk 4]
```diff
@@ -7,8 +7,11 @@ SPDX-License-Identifier: Apache-2.0
 package ttx
 
 import (
+	"encoding/base64"
 	"strconv"
 
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/services/comm"
+
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/services/ttx/support.go]: done

File: integration/token/fungible/views/issue.go [hunk 5]
```diff
@@ -3,29 +3,28 @@ Copyright IBM Corp. All Rights Reserved.
...
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/network"
 	"github.com/hyperledger-labs/fabric-token-sdk/token/services/ttx"
+	token2 "github.com/hyperledger-labs/fabric-token-sdk/token/token"
 )
 
 // IssueCash contains the input information to issue a token
 type IssueCash struct {
 	// Anonymous set to true if the transaction is anonymous, false otherwise
 	Anonymous bool
+	// Auditor is the name of the auditor identity
```
Comment (KElkhiyaoui) [inline integration/token/fungible/views/issue.go:26]: is it name or enrolment ID?

File: integration/token/fungible/views/redeem.go [hunk 6]
```diff
@@ -21,6 +22,8 @@ import (
 
 // Redeem contains the input information for a redeem
 type Redeem struct {
+	// Auditor is the name of the auditor that must be contacted to approve the operation
```
Comment (KElkhiyaoui) [inline integration/token/fungible/views/redeem.go:25]: name vs. enrolment ID?

File: integration/token/fungible/views/swap.go [hunk 7]
```diff
@@ -20,6 +21,8 @@ import (
 
 // Swap contains the input information for a swap
 type Swap struct {
+	// Auditor is the identity of the auditor that must be contacted to approve the transaction
```
Comment (KElkhiyaoui) [inline integration/token/fungible/views/swap.go:24]: is it identity or enrolment ID?

File: integration/token/fungible/views/transfer.go [hunk 8]
```diff
@@ -24,6 +25,8 @@ import (
 
 // Transfer contains the input information for a transfer
 type Transfer struct {
+	// Auditor is the name of the auditor that must be contacted to approve the operation
```
Comment (KElkhiyaoui) [inline integration/token/fungible/views/transfer.go:28]: name vs enrolment ID?

**Review by mbrandenburger — APPROVED** [review]

File: token/services/ttx/support.go [hunk 9]
```diff
@@ -54,3 +57,128 @@ func StoreEnvelope(context view.Context, tx *Transaction) error {
...
 }
+
+// RunView runs passed view within the passed context and using the passed options in a separate goroutine
+func RunView(context view.Context, view view.View, opts ...view.RunViewOption) {
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Debugf("panic in RunView: %v", r)
+		}
+	}()
+	go func() {
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply token/services/ttx/support.go:65]: I've created hyperledger-labs/fabric-smart-client#313

File: token/sdk/network/ownership.go [hunk 10]
```diff
@@ -13,6 +13,7 @@ import (
 
 type Ownership interface {
```
Comment (KElkhiyaoui) [inline token/sdk/network/ownership.go:14]: shall we change the name of this interface to Authorization?

File: integration/token/fungible/views/issue.go (hunk 5 above)
Comment (adecaro) [reply integration/token/fungible/views/issue.go:26]: it is just the name/label in the configuration used to identify the FSC auditor identity.

**Review by KElkhiyaoui — COMMENTED** [review]
some minor refactoring is required.

File: integration/token/fungible/views/redeem.go (hunk 6 above)
Comment (adecaro) [reply integration/token/fungible/views/redeem.go:25]: not the EID but just the name/label used in the configuration to identify the FSC auditor identity

File: integration/token/fungible/views/transfer.go (hunk 8 above)
Comment (adecaro) [reply integration/token/fungible/views/transfer.go:28]: not the EID but just the name/label used in the configuration to identify the FSC auditor identity

File: token/sdk/network/ownership.go (hunk 10 above)
Comment (adecaro) [reply token/sdk/network/ownership.go:14]: okay

**Comment by adecaro** [comment]
Thanks @KElkhiyaoui , I have addressed your comments.

File: token/sdk/network/authrorization.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	return false
+}
+
+// AuthorizationMultiplexer iterates over multiple authorization checker
+type AuthorizationMultiplexer struct {
+	ownerships []Authorization
+}
+
+// NewOwnershipMultiplexer returns a new AuthorizationMultiplexer for the passed ownership checkers
+func NewOwnershipMultiplexer(ownerships ...Authorization) *AuthorizationMultiplexer {
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/sdk/network/authrorization.go:53]: good catch

[the remark this answers is not in the document]
Comment (adecaro) [reply token/sdk/network/authrorization.go]: done

**Review by KElkhiyaoui — APPROVED** [review]

**Pull request merged**